src/lb_data.c

```diff
@@ -1241,10 +1494,29 @@ int lb_halo_post(const lb_t * lb, lb_halo_t * h) {
 
   TIMER_start(TIMER_LB_HALO_PACK);
 
-  #pragma omp parallel
-  {
-    for (int ireq = 0; ireq < h->map.nvel; ireq++) {
-      lb_halo_enqueue_send(lb, h, ireq);
+  int ndevice;
+  tdpGetDeviceCount(&ndevice);
+  if (ndevice > 0 && lb->haloscheme == LB_HALO_TARGET) {
+    if (have_graph_api_) {
+      tdpAssert( tdpGraphLaunch(h->gsend.exec, h->stream) );
+      tdpAssert( tdpStreamSynchronize(h->stream) );
+    } else {
+      for (int ireq = 0; ireq < h->map.nvel; ireq++) {
+        if (h->count[ireq] > 0) {
+          int scount = h->count[ireq]*lb_halo_size(h->slim[ireq]);
+          dim3 nblk, ntpb;
+          kernel_launch_param(scount, &nblk, &ntpb);
+          tdpLaunchKernel(lb_halo_enqueue_send_kernel, nblk, ntpb, 0, 0, lb->target, h->target, ireq);
+          tdpDeviceSynchronize();
+        }
```

Comment:
More importantly, in the case where gpu-aware MPI is not available, the kernel needs to be followed by the corresponding `memcpy()`.